File: rust/examples/basic_example.rs
// Copyright (c) Microsoft Corporation. All rights reserved.

//! Basic example demonstrating the Copilot SDK for Rust.
//!
//! This example shows how to:
//! - Create and start a client, then connect to the CLI server
//! - Subscribe to session events
//! - Send messages and wait for responses

use copilot_sdk::*;

#[tokio::main]
async fn main() -> Result<(), CopilotError> {
    // ========================================================================
    // 1. Create and start the client
    // ========================================================================
    let client = CopilotClient::new(CopilotClientOptions::default());

    println!("Starting Copilot client...");
    client.start().await?;
    println!("Client connected!");

    // Verify connectivity with a ping
    let ping = client.ping(Some("hello from Rust SDK")).await?;
    println!(
        "Ping response: {} (protocol v{})",
        ping.message,
        ping.protocol_version.unwrap_or(0)
    );

    // ========================================================================
    // 2. List available models (optional)
    // ========================================================================
    match client.list_models().await {
        Ok(models) => {
            println!("\nAvailable models:");
            for model in &models {
                println!("  - {} ({})", model.name, model.id);
            }
        }
        Err(e) => println!("Could not list models: {}", e),
    }

    // ========================================================================
    // 3. Create a session
    // ========================================================================
    let session = client.create_session(SessionConfig::default()).await?;
    println!("\nSession created: {}", session.session_id());

    // ========================================================================
    // 4. Subscribe to events
    // ========================================================================
    let _sub = session
        .on(|event| {
            if let Some(content) = event.assistant_message_content() {
                print!("{}", content);
            }
        })
        .await;

    // ========================================================================
    // 5. Send a message and wait for the response
    // ========================================================================
    println!("\nSending message...");
    let response = session
        .send_and_wait(
            MessageOptions {
                prompt: "What is the weather in San Francisco?".to_string(),
                ..Default::default()
            },
            Some(120_000),
        )
        .await?;

    match response {
        Some(event) => {
            if let Some(content) = event.assistant_message_content() {
                println!("\nsend_and_wait response: {}", content);
            }
        }
        None => println!("\nNo assistant message received."),
    }

    // ========================================================================
    // 6. Send a follow-up message
    // ========================================================================
    println!("\nSending follow-up...");
    let response = session
        .send_and_wait(
            MessageOptions {
                prompt: "Now what about New York?".to_string(),
                ..Default::default()
            },
            None,
        )
        .await?;

    if let Some(event) = response {
        if let Some(content) = event.assistant_message_content() {
            println!("\nFollow-up response: {}", content);
        }
    }

    // ========================================================================
    // 7. Session metadata (optional)
    // ========================================================================
    if let Ok(meta) = client.get_session_metadata(session.session_id()).await {
        println!("\nSession metadata: {:?}", meta);
    }

    // ========================================================================
    // 8. Clean up
    // ========================================================================
    println!("\nCleaning up...");
    session.destroy().await?;
    if let Err(e) = client.stop().await {
        eprintln!("Cleanup errors: {}", e);
    }

    println!("Done!");
    Ok(())
}

File: scala/examples/BasicExample.scala
// Copyright (c) Microsoft Corporation. All rights reserved.

// Basic example demonstrating the Copilot Scala SDK.
// Run with: sbt "runMain BasicExample"

import com.github.copilot.*
import io.circe.*
import io.circe.syntax.*

import scala.concurrent.{Await, ExecutionContext, Future}
import scala.concurrent.duration.*

@main def BasicExample(): Unit =
  given ExecutionContext = ExecutionContext.global

  // -------------------------------------------------------------------------
  // 1. Create the client
  // -------------------------------------------------------------------------
  // Default options spawn the bundled CLI process using stdio transport.
  // Set cliPath if the CLI is at a custom location.
  val client = CopilotClient(CopilotClientOptions(
    cliPath = Some("copilot"),  // adjust to your CLI path
    logLevel = "info"
  ))

  try
    // Start the client (spawns CLI process, connects via stdio, verifies protocol)
    println("Starting Copilot client...")
    Await.result(client.start(), 30.seconds)
    println(s"Client connected (state: ${client.getState})")

    // -----------------------------------------------------------------------
    // 2. Ping the server
    // -----------------------------------------------------------------------
    val pingResponse = Await.result(client.ping(Some("hello from Scala")), 10.seconds)
    println(s"Ping response: message=${pingResponse.message}, " +
      s"protocolVersion=${pingResponse.protocolVersion.getOrElse("unknown")}")

    // -----------------------------------------------------------------------
    // 3. List available models
    // -----------------------------------------------------------------------
    val models = Await.result(client.listModels(), 15.seconds)
    println(s"Available models (${models.size}):")
    models.take(5).foreach { m =>
      println(s"  - ${m.id} (${m.name})")
    }

    // -----------------------------------------------------------------------
    // 4. Define a custom tool
    // -----------------------------------------------------------------------
    val weatherTool = DefineTool(
      name = "get_weather",
      description = "Get the current weather for a location",
      parameters = JsonObject(
        "type" -> "object".asJson,
        "properties" -> Json.obj(
          "location" -> Json.obj(
            "type" -> "string".asJson,
            "description" -> "The city and state, e.g. San Francisco, CA".asJson
          )
        ),
        "required" -> Json.arr("location".asJson)
      )
    ) { (args, invocation) =>
      val location = args.hcursor.get[String]("location").getOrElse("unknown")
      println(s"  [Tool] get_weather called for: $location")
      Future.successful(ToolResultObject(
        textResultForLlm = s"The weather in $location is 72F and sunny.",
        resultType = ToolResultType.Success,
        toolTelemetry = Some(Map.empty)
      ))
    }

    // -----------------------------------------------------------------------
    // 5. Create a session with the tool
    // -----------------------------------------------------------------------
    println("\nCreating session...")
    val session = Await.result(
      client.createSession(SessionConfig(
        tools = List(weatherTool),
        onPermissionRequest = Some { (request, invocation) =>
          println(s"  [Permission] ${request.kind} requested")
          Future.successful(PermissionRequestResult(kind = PermissionRequestResult.Approved))
        }
      )),
      30.seconds
    )
    println(s"Session created: ${session.sessionId}")

    // -----------------------------------------------------------------------
    // 6. Subscribe to events
    // -----------------------------------------------------------------------
    session.on { event =>
      event.`type` match
        case SessionEventType.AssistantMessage =>
          val content = event.data.hcursor.get[String]("content").getOrElse("")
          println(s"\n  [Assistant] $content")
        case SessionEventType.ToolExecutionStart =>
          val toolName = event.data.hcursor.get[String]("toolName").getOrElse("")
          println(s"  [Tool Start] $toolName")
        case SessionEventType.ToolExecutionComplete =>
          val toolName = event.data.hcursor.get[String]("toolName").getOrElse("")
          println(s"  [Tool Complete] $toolName")
        case SessionEventType.SessionIdle =>
          println("  [Session Idle]")
        case SessionEventType.SessionError =>
          val msg = event.data.hcursor.get[String]("message").getOrElse("unknown error")
          println(s"  [Error] $msg")
        case _ =>
          () // ignore other events
    }

    // -----------------------------------------------------------------------
    // 7. Send a message and wait for the response
    // -----------------------------------------------------------------------
    println("\nSending message...")
    val response = Await.result(
      session.sendAndWait(MessageOptions(
        prompt = "What is the weather in San Francisco?"
      )),
      120.seconds
    )

    response match
      case Some(event) =>
        val content = event.data.hcursor.get[String]("content").getOrElse("")
        println(s"\nFinal response: $content")
      case None =>
        println("\nNo assistant message received.")

    // -----------------------------------------------------------------------
    // 8. Send another message (conversation continues)
    // -----------------------------------------------------------------------
    println("\nSending follow-up message...")
    val followUp = Await.result(
      session.sendAndWait(MessageOptions(
        prompt = "What about New York?"
      )),
      120.seconds
    )

    followUp match
      case Some(event) =>
        val content = event.data.hcursor.get[String]("content").getOrElse("")
        println(s"\nFollow-up response: $content")
      case None =>
        println("\nNo follow-up response received.")

    // -----------------------------------------------------------------------
    // 9. Get message history
    // -----------------------------------------------------------------------
    val messages = Await.result(session.getMessages(), 10.seconds)
    println(s"\nSession has ${messages.size} events in history.")

    // --- v2.0 Features ---

    // Session Metadata
    val meta = Await.result(client.getSessionMetadata(session.sessionId), 10.seconds)
    meta.foreach(m => println(s"Session model: ${m.model}"))

    // Skills (uncomment to use)
    // val skillSession = Await.result(client.createSession(SessionConfig(
    //   skillDirectories = Seq("./skills"),
    //   includeSubAgentStreamingEvents = Some(true)
    // )), 30.seconds)

    // -----------------------------------------------------------------------
    // 10. Clean up
    // -----------------------------------------------------------------------
    println("\nDestroying session...")
    Await.result(session.destroy(), 10.seconds)
    println("Session destroyed.")

  finally
    println("Stopping client...")
    val errors = client.stop()
    if errors.nonEmpty then
      println(s"Cleanup errors: ${errors.map(_.getMessage).mkString(", ")}")
    else
      println("Client stopped cleanly.")
